src/librbd/ManagedLock.cc

```diff
@@ -762,8 +762,9 @@ void ManagedLock<I>::handle_shutdown_pre_release(int r) {
   using managed_lock::ReleaseRequest;
   ReleaseRequest<I>* req = ReleaseRequest<I>::create(m_ioctx, m_watcher,
       m_work_queue, m_oid, cookie,
-      new FunctionContext([this](int r) {
-        post_release_lock_handler(true, r, create_context_callback<
+      new FunctionContext([this, r](int l) {
+        int rst = r < 0 ? r : l;
```

Comment:
That is where bug lies. The r guaranteed to 0 causes this problem:  In PreReleaseRequest.cc `io_work_queue->unblock_writes` will be called as in `handle_invalidate_cache when r < 0`, `handle_block_writes when r <0`，and in ExclusiveLock.cc `post_release_lock_handler` when r >= 0 and shutting_down. Because the r in post_release_lock_handler is guaranteed to be 0 so when handle_block_writes fails, `unblock_writes` is called twice. So we have to pass the status in PreReleaseRequest.cc out to post_release_lock_handler, so it will not call unblock_writes.

Reply:
As to ` an ASOK command to purge the in-memory cache`,  I think this pr basically fix two things: 

1. When  some io ops fail, unblock_writes will not be called twice (which cause assert failure). 2. Allow clean shutting down when io ops fail because of -EIO. 

` an ASOK command to purge the in-memory cache` may not work. See this case: `call purge cache` -> new aio arrives, fill cache -> detach disk, thus close image -> `cancel all aio` -> block_writes fails. It is hard to guarantee cache is empty before detach disk, so it is better to let PreReleaseRequest shutdown cleanly when some ops fail.

Reply:
@liyichao I am still not comfortable putting these intrusive changes in to support your non-standard use-case. Using your example of incoming IO, if you release the lock and unblock the writes, you have now potentially corrupted the image.

Reply:
OK, then. I will remove -EIO related changes. How about the first case `When some io ops fail, unblock_writes will not be called twice (which cause assert failure). `, it is a bug that happens when aio fail, for example, you set `rados osd time out`, and it times out, unblock_writes will be called twice.

The bug related changes are only this line `int rst = r < 0 ? r : l;`.

Reply:
That's sounds good to me. Thanks.